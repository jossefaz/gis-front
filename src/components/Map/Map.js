import React from "react";
import PropTypes from "prop-types";
import API from "../../core/api";
import VectorLayerRegistry from "../../core/proxymanagers/vectorlayer";
import _ from "lodash";

const { getFocusedMap, getFocusedMapUUID } = API.map;
class MapComponent extends React.Component {
  state = {};

  defaultClickTool = async (e) => {
    const opennedTools = this.props.Tools[getFocusedMapUUID()];
    if (
      opennedTools.dynamicTools.length === 0 &&
      Object.keys(this.props.CurrentInteractions).length === 0
    ) {
      const features = this.vectorLayerRegistry.getFeaturesAtCoordinate(
        e.coordinate
      );
      if (features.length > 0) {
        await this.props.setSelectedFeatures(features);
        await this.props.toggleToolByName("Identify", true);
      }
    }
  };
  get vectorLayerRegistry() {
    return VectorLayerRegistry.getInstance();
  }

  componentDidMount() {
    if (getFocusedMap().getLayers().getArray().length == 0) {
      this.props.setRaster("osm");
    }
    getFocusedMap().on("pointerdown", this.defaultClickTool);
    if (getFocusedMapUUID() in this.props.Layers) {
      this.setState({
        currentLayers: [],
      });
    }
  }
  componentDidUpdate() {
    if (getFocusedMap().getLayers().getArray().length == 0) {
      this.props.setRaster("osm");
    }
    const areEquals = _.isEqual(
      this.state.currentLayers,
      this.props.VisibleLayers
    );
    if (!areEquals) {
      this.vectorLayerRegistry.initVectorLayers(this.props.VisibleLayers);
      this.setState({
        currentLayers: this.props.VisibleLayers,
      });
    }
  }
  render() {
    return <div id="map" className="map"></div>;
  }
}

MapComponent.propTypes = {
  setRaster: PropTypes.func, // should be imported from redux action
};

export default MapComponent;
