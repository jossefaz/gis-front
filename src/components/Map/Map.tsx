import React, { FC, useEffect, useState } from "react";
import PropTypes from "prop-types";
import API from "../../core/api";
import VectorLayerRegistry from "../../core/proxymanagers/vectorlayer";
import _ from "lodash";
import { InteractionUtil } from "../../utils/interactions";
import { useTypedSelector } from "../../hooks/useTypedSelectors";
import {
  selectVisibleLayers,
  selectCurrentInteractions,
} from "../../state/reducers";
import { useActions } from "../../hooks/useActions";
import { shiftKeyOnly } from "ol/events/condition";
import { DragBox } from "ol/interaction";
import { boundingExtent, buffer } from "ol/extent";
import { MapBrowserEvent } from "ol";

const { getFocusedMap, getFocusedMapUUID } = API.map;
const MapComponent: FC = () => {
  const WIDGET_NAME = "Map";
  const interactions = new InteractionUtil(WIDGET_NAME);
  const vlregistry = VectorLayerRegistry.getInstance();
  const storeTools = useTypedSelector(
    (state) => state.Tools[getFocusedMapUUID()]
  );
  const visibleLayers = useTypedSelector(selectVisibleLayers);
  const currentInteraction = useTypedSelector(selectCurrentInteractions);
  const { setSelectedFeatures, toggleToolByName, setRaster } = useActions();
  const [currentLayers, setcurrentLayers] = useState<string[]>([]);
  const openedTools =
    storeTools.stickyTool.length > 0 ||
    storeTools.dynamicTools.length > 0 ||
    Object.keys(currentInteraction).length > 0;

  const onBoxEnd = () => {
    if (interactions.currentDragBoxUUID) {
      const dragBox = interactions.currentDragBox;
      const endListener = (dragBox: DragBox) => {
        const extent = dragBox.getGeometry().getExtent();
        const features = vlregistry.getFeaturesByExtent(extent);
        if (Object.keys(features).length > 0) {
          setSelectedFeatures(features);
          toggleToolByName("Identify", true, false);
        }
      };
      if (dragBox) {
        dragBox.on("boxend", () => endListener(dragBox));
      }
    }
  };

  const defaultClickTool = async (e: MapBrowserEvent<UIEvent>) => {
    if (!openedTools && Object.keys(currentInteraction).length === 0) {
      interactions.newDragBox(shiftKeyOnly);
      onBoxEnd();
      const extent = buffer(boundingExtent([e.coordinate]), 10);
      const features = vlregistry.getFeaturesByExtent(extent);
      if (Object.keys(features).length > 0) {
        setSelectedFeatures(features);
        toggleToolByName("Identify", true, false);
      }
    } else {
      interactions.unDragBox();
    }
  };
  const areLayersEquals = _.isEqual(currentLayers, visibleLayers);

  useEffect(() => {
    if (getFocusedMap().getLayers().getArray().length == 0) {
      setRaster("osm");
    }
    if (!areLayersEquals) {
      vlregistry.initVectorLayers(visibleLayers);
      setcurrentLayers(visibleLayers);
    }
    if (!openedTools) {
      getFocusedMap().on("click", defaultClickTool);
    }
  });

  useEffect(() => {
    return () => {
      interactions.unDragBox();
    };
  }, []);

  return <div id="map" className="map"></div>;
};

export default MapComponent;
