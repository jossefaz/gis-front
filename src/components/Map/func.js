import { Image as ImageLayer } from "ol/layer";
import axios from "axios";




export const Identify = (evt, mapObject, actionCB) => {
  var viewResolution = mapObject.getView().getResolution();
  mapObject
    .getLayers()
    .getArray()
    .map((lyr) => {
      if (lyr instanceof ImageLayer && lyr.selectable) {
        var url = lyr
          .getSource()
          .getFeatureInfoUrl(evt.coordinate, viewResolution, "EPSG:4326", {
            INFO_FORMAT: "application/json",
            feature_count: 100,
          });
        if (url) {
          axios.get(url).then((response) => {
            actionCB(response.data.features);
          });
        }
      }
    });
};

