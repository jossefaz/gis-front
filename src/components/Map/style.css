.map {
  width: 100%;
  height: 95.5vh;
  margin-left: 1em;
  background-color: "#cccccc";
}
