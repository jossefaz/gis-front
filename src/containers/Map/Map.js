import React from "react";
import { connect } from "react-redux";
import { setSelectedFeatures } from "../../redux/actions/features";
import { setRaster } from "../../redux/actions/raster";
import MapComp from "../../components/Map/Map";
import { getFocusedMap } from "../../nessMapping/api";
class MapComponent extends React.Component {
  render() {
    return <MapComp getFocusedMap={getFocusedMap} {...this.props} />;
  }
}
const mapStateToProps = (state) => {
  return {
    map: state.map,
  };
};

export default connect(mapStateToProps, { setSelectedFeatures, setRaster })(
  MapComponent
);
