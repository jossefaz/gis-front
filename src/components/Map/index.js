import React from "react";
import { InitMap } from '../../utils/Map'
import config from 'react-global-configuration';
import {logLevel, LogIt} from '../../utils/logs'
import './style.css'
class MapComponent extends React.Component {
  constructor(props) {
        super(props);       
        this.map = {};
    }
  componentDidMount() {
    this.map = InitMap();
    LogIt(logLevel.INFO, "Map init" )
    LogIt(logLevel.DEBUG, this.map )
  }
  render() {
    const {target} = config.get("MapConfig")
    return <div id={target} className="map"></div>;
  }
}
export default MapComponent;