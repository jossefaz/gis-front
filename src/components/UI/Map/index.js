import React from "react";
import { InitMap } from "./func";
import { connect } from "react-redux";
import config from "react-global-configuration";
import { logLevel, LogIt } from "../../../utils/logs";
import { InitLayers } from "../../../redux/actions/layers";
import "./style.css";
class MapComponent extends React.Component {
  constructor(props) {
    super(props);
    this.map = null;
  }
  componentDidMount() {
    this.map = InitMap();
    LogIt(logLevel.INFO, "Map init");
    LogIt(logLevel.DEBUG, this.map);
    this.props.InitLayers(config.get("layers"));
  }

  componentDidUpdate() {
    console.log("MAP UPDATE");
    Object.keys(this.props.mapLayers).map((lyrId) =>
      this.map.addLayer(this.props.mapLayers[lyrId])
    );
  }

  render() {
    const { target } = config.get("MapConfig");
    return <div id={target} className="map"></div>;
  }
}
const mapStateToProps = (state) => {
  return { mapLayers: state.mapLayers };
};

export default connect(mapStateToProps, { InitLayers })(MapComponent);
