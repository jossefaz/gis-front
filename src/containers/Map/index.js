import React from "react";
import { InitMap, addLayersSafely } from "./func";
import { connect } from "react-redux";
import config from "react-global-configuration";
import { logLevel, LogIt } from "../../utils/logs";
import { InitLayers, addLayers } from "../../redux/actions/layers";
import { InitRasters } from "../../redux/actions/raster";
import "./style.css";
class MapComponent extends React.Component {
  constructor(props) {
    super(props);
    this.map = null;
  }
  componentDidMount() {
    this.map = InitMap();
    LogIt(logLevel.INFO, "Map init");
    LogIt(logLevel.DEBUG, this.map);
    this.props.InitLayers(config.get("layers"));
    this.props.InitRasters();
  }

  addLayersSafely = (layers) => {
    const addedToMap = [];
    Object.keys(layers).map((lyrId) => {
      if (!layers[lyrId].addedToMap) {
        this.map.addLayer(layers[lyrId]);
        addedToMap.push(lyrId);
      }
    });
    if (addedToMap.length > 0) {
      this.props.addLayers(addedToMap);
    }
  };

  componentDidUpdate() {
    LogIt(logLevel.INFO, "Map Update");
    LogIt(logLevel.DEBUG, this.props.Layers);
    if (this.props.Rasters) {
      const { Catalog, Focused } = this.props.Rasters;
      this.map.getLayers().setAt(0, Catalog[Focused].layer);
    }
    this.addLayersSafely(this.props.Layers);
  }

  render() {
    const { target } = config.get("MapConfig");
    return <div id={target} className="map"></div>;
  }
}
const mapStateToProps = (state) => {
  return { Layers: state.mapLayers, Rasters: state.Rasters };
};

export default connect(mapStateToProps, { InitLayers, InitRasters, addLayers })(
  MapComponent
);
