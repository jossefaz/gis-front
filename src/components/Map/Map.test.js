import { shallow } from "enzyme";
import React from "react";
import Map from "./Map";

let wrapper;
beforeEach(() => {
  const mockProps = {
    setRaster: jest.fn(),
    map: {
      getLayers: jest.fn(() => {
        return {
          getArray: jest.fn(() => {
            return [1];
          }),
        };
      }),
    },
    getFocusedMap: jest.fn(() => {
      return {
        getLayers: jest.fn(() => {
          return {
            getArray: jest.fn(() => {
              return [1];
            }),
          };
        }),
      };
    }),
  };
  wrapper = shallow(<Map {...mockProps} />);
});

describe("Map Component", () => {
  it("render Map Component without crashing", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
