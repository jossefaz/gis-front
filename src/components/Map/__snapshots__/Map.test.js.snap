// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map Component render Map Component without crashing 1`] = `
<div
  className="map"
  id="map"
/>
`;
